File: DesignPatterns/01_GOF23/DP_examples/Python/03_Behavioral/TemplateMethod/ConcreteClass.py
# 具体饮料实现
from AbstractClass import BeverageMaker

# 咖啡制作类
class CoffeeMaker(BeverageMaker):
    # 实现冲泡方法
    def brew(self):
        print("Brewing coffee grounds")
    
    # 实现添加调料方法
    def add_condiments(self):
        print("Adding sugar and milk")
    
    # 重写钩子方法
    def customer_wants_condiments(self):
        print("Customer wants sugar and milk")
        return True

# 茶制作类
class TeaMaker(BeverageMaker):
    # 实现冲泡方法
    def brew(self):
        print("Steeping the tea bag")
    
    # 实现添加调料方法
    def add_condiments(self):
        print("Adding lemon")
    
    # 重写钩子方法
    def customer_wants_condiments(self):
        print("Customer wants lemon")
        return True

File: DesignPatterns/01_GOF23/DP_examples/Lua/03_Behavioral/TemplateMethod/ConcreteClass.lua
-- 具体饮料实现
local BeverageMaker = require("AbstractClass")

-- 咖啡制作类
local CoffeeMaker = BeverageMaker:new()

function CoffeeMaker:brew()
    print("Brewing coffee grounds")
end

function CoffeeMaker:addCondiments()
    print("Adding sugar and milk")
end

function CoffeeMaker:customerWantsCondiments()
    print("Customer wants sugar and milk")
    return true
end

-- 茶制作类
local TeaMaker = BeverageMaker:new()

function TeaMaker:brew()
    print("Steeping the tea bag")
end

function TeaMaker:addCondiments()
    print("Adding lemon")
end

function TeaMaker:customerWantsCondiments()
    print("Customer wants lemon")
    return true
end

return {
    CoffeeMaker = CoffeeMaker,
    TeaMaker = TeaMaker
}
